aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -570,6 +574,133 @@ struct cpu_scatter_gather_base_kernel {
   }
 };
 
+// Note [scatter reduce optimization]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//
+// 1. initiative: optimize scatter_reduce optimization on PyG
+//   `scatter_add` is extensively used on 'message passing' when
+//   aggregating info. The `index` tensor is extended which means
+//   the aggregation is on rowwise.
+//
+// 2. implementation: map `scatter_reduce` to `spmm` reduce
+//   in the shape of `[M, N]` * `[N, K]`, where:
+//
+//   M: self_dim_size
+//   nnz: index_dim_size
+//   K: index.numel() / index_dim_size;
+//
+//   step 1: convert input index to CSR format (use radix_sort to
+//     solve write addr conflicts on `self` tensor)
+//
+//   step 2: spmm reduce, parallel on M and vectorize on K
+//
+template <typename scalar_t>
+void cpu_scatter_add_contig_kernel(const Tensor& self, const Tensor& index, const Tensor& src) {
+  int64_t* index_data = index.data_ptr<int64_t>();
+  scalar_t* self_data = self.data_ptr<scalar_t>();
+  scalar_t* src_data = src.data_ptr<scalar_t>();
+
+  const int64_t M = ensure_nonempty_size(self, 0);
+  const int64_t nnz = ensure_nonempty_size(index, 0);
+  const int64_t K = index.numel() / nnz;
+
+  const int64_t index_upper_bound = M;
+
+  std::unique_ptr<int64_t []> keys(new int64_t[nnz]);
```

Comment:
sorry i don't quite understand here. Are you referring to the MACRO to get `index` pointers?

Reply:
@rusty1s Do you mean something like `_convert_indices_from_coo_to_csr`? I think this function assume the indices are sorted while here we cannot assume this and have to do sorting efficiently before converting to csr.

Reply:
Yeah, that's what I meant. We can still leverage it after the radix sort as far as I can tell.

Reply:
oh I see, will update this :)

Reply:
I changed the code using `_convert_indices_from_coo_to_csr` and then go to `addmm`

```
Tensor crow_indices_tensor = at::_convert_indices_from_coo_to_csr(
       sorted_col_index_keys_tensor,
       M,
       /*out_int32*/ false);

   Tensor col_indices_tensor = at::from_blob(
       (void*)sorted_col_index_values,
       {nnz},
       self.options().dtype(at::kLong).device(at::kCPU));

   Tensor sparse_values = at::ones({nnz}, self.options());
   Tensor sparse = at::sparse_csr_tensor(crow_indices_tensor, col_indices_tensor, sparse_values, c10::CppTypeToScalarType<scalar_t>::value);

   at::addmm_out(const_cast<Tensor&>(self), self, sparse, src);
```

This would be about 1.8x slower compared to our previous result, the reason is the `row_indices` in this PR is not entirely the same as standard `CSR`, only the nonzero rows for `self` are recorded so the nodes without any connections will not be reflected in the parallelization (won't make CPU cores run idle). This is the same technique we use when optimizing `EmbeddingBag`.

I think it's better to keep the current implementation since it is actually a more performant algorithm.